File: contracts/GlassLink.kt
package com.ambientlink.contracts

import kotlinx.coroutines.flow.StateFlow

/**
 * CANONICAL CONTRACT — copy into each vendor repo and implement against the real
 * device transport. Do not add vendor types here.
 *
 * Extracted from the recovered Cosmo `CosmoGlassManager` shape
 * (see ambient-link-google/glasses_link.md). The transport below this interface is
 * always vendor-hidden; this is the stable boundary above it.
 *
 * Implementations MUST:
 *  - expose link state only as StateFlow (no polled booleans),
 *  - make bind() idempotent (no-op if bound or binding in flight),
 *  - push media through callbacks (frames/audio), never block,
 *  - throttle frames (default 1 / 10s) and route media through an EphemeralBuffer,
 *  - run capture inside a typed foreground service,
 *  - honor a settings gate before binding.
 */
interface GlassLink {
    /** Device present/reachable (e.g. ProjectedContext.isProjectedDeviceConnected). */
    val connected: StateFlow<Boolean>

    /** Capture service bound / session live. */
    val bound: StateFlow<Boolean>

    /** Idempotent. Honors the per-link settings gate; no-op when already (un)bound. */
    suspend fun bind()
    fun unbind()

    /** Register the frame sink before starting image capture. */
    fun setupImageCapture(onFrame: (Frame) -> Unit)
    fun startImageCapture()
    fun stopImageCapture()

    /** Audio bytes + valid length, fed into the shared STT pipeline. */
    fun startAudioCapture(onBytes: (ByteArray, Int) -> Unit)
    fun stopAudioCapture()

    /** Drop buffers and reset all gates. */
    fun clear()

    /** Vendor-neutral frame envelope; impls wrap their native bitmap/proxy. */
    data class Frame(val width: Int, val height: Int, val pixels: ByteArray, val tsMillis: Long)

    companion object {
        /** Cosmo: FRAME_PROCESS_INTERVAL_MS = 10_000 ms, GLASS_CAMERA_TARGET_FPS = 0.1. */
        const val DEFAULT_FRAME_INTERVAL_MS: Long = 10_000
    }
}

/**
 * TTL ring buffer for captured media. Mirrors Cosmo's InMemoryEphemeralBuffer +
 * getEphemeralBufferDurationMin: bounded, time-expiring, clearable. Keeps memory
 * flat under continuous capture.
 */
interface EphemeralBuffer<T> {
    fun add(item: T, tsMillis: Long)
    fun snapshot(): List<T>
    fun clear()
    /** Items older than this are evicted on add()/snapshot(). */
    val ttlMillis: Long
}

File: core-apple/Sources/AmbientLinkCore/GlassLink.swift
import Foundation

/// Vendor-neutral companion-capture contract for Apple platforms. Canonical here;
/// ambient-link-apple and ambient-link-meta/relay-ios depend on this instead of
/// carrying copies.
///
/// visionOS has no projected capture service like Cosmo's, but the boundary still
/// applies: a GlassLink impl can wrap the audio engine + relay so the Siri/App
/// Intents layer and the RealityKit frontend consume one uniform surface. Shape
/// extracted from the recovered Cosmo CosmoGlassManager
/// (ambient-link-google/glasses_link.md); plan in ambient-link-core/ROUTING.md.
public struct GlassFrame: Sendable {
    public let width: Int
    public let height: Int
    public let pixels: Data
    public let timestamp: Date
    public init(width: Int, height: Int, pixels: Data, timestamp: Date) {
        self.width = width; self.height = height; self.pixels = pixels; self.timestamp = timestamp
    }
}

public protocol GlassLink: AnyObject {
    /// Device present/reachable.
    var connected: AsyncStream<Bool> { get }
    /// Capture session live.
    var bound: AsyncStream<Bool> { get }

    /// Idempotent; honors the per-link settings gate.
    func bind() async
    func unbind()

    func setupImageCapture(onFrame: @escaping @Sendable (GlassFrame) -> Void)
    func startImageCapture()
    func stopImageCapture()

    /// Audio bytes + valid length, fed into the shared STT pipeline.
    func startAudioCapture(onBytes: @escaping @Sendable (Data, Int) -> Void)
    func stopAudioCapture()

    func clear()
}

public enum GlassLinkDefaults {
    /// Cosmo: 10s frame interval / 0.1 fps target.
    public static let frameIntervalMillis: Int = 10_000
}
